src/transformStateWithClones.js

```diff
@@ -7,7 +7,42 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = []; // will be our result
+  let currentState = { ...state }; /* create copy of the
+  initial object so later we can push more copies of this object
+   to the stateHistory array at each iteration cycle without modifying
+   initial state object */
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        currentState = { ...currentState, ...action.extraData };
+        /* if action array has an object with 'type: addProperties' property,
+        we add properties from extraData object and create another copy of
+        current state */
+        break;
+      }
+
+      case 'removeProperties': {
+        currentState = { ...currentState };
```

Comment:
[DONT REPEAT YOURSELF]: There is no need to create a new copy of `currentState` in the 'removeProperties' case since you are already creating a copy at the beginning of each iteration. You can remove this line to avoid unnecessary object copying.